### PR TITLE
fix: redis connect with promise race

### PR DESCRIPTION
**Pull request by koladilip**
## Description of the change

timeoutPromise should be a function so that it returns a new promise everytime.

## Type of change
- [x] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

**Review by anantjain45823 — DISMISSED** [review]

**Review by yashasvibajpai — APPROVED** [review]

**Review by anantjain45823 — APPROVED** [review]

File: src/util/redisConnector.js [hunk 1]
```diff
@@ -3,10 +3,10 @@ const { RedisError } = require('../v0/util/errorTypes');
 const log = require('../logger');
 const stats = require('./stats');
 
-const timeoutPromise = new Promise((resolve, reject) => {
```
Comment (krishna2020) [inline src/util/redisConnector.js:6]: @anantjain45823 check out this code fix & test out in local on the connection scenarios

File: src/v0/sources/shopify/transform.js [hunk 2]
```diff
@@ -178,7 +179,12 @@ const isIdentifierEvent = (event) => {
 const processIdentifierEvent = async (event, metricMetadata) => {
   if (useRedisDatabase) {
     const setStartTime = Date.now();
-    await RedisDB.setVal(`${event.cartToken}`, { anonymousId: event.anonymousId });
+    try {
+      await RedisDB.setVal(`${event.cartToken}`, { anonymousId: event.anonymousId }
+      );
```
Comment (koladilip) [inline src/v0/sources/shopify/transform.js:184]: new line required?

File: src/v0/sources/shopify/transform.js [hunk 3]
```diff
@@ -178,7 +179,12 @@ const isIdentifierEvent = (event) => {
...
+    } catch (e) {
+      // This is a client side event
```
Comment (koladilip) [inline src/v0/sources/shopify/transform.js:186]: is there metric for this already?

File: src/v0/sources/shopify/transform.js [hunk 4]
```diff
@@ -196,7 +202,8 @@ const processIdentifierEvent = async (event, metricMetadata) => {
   };
   return result;
 };
-const process = async (event) => {
+const process1 = async (event) => {
```
Comment (koladilip) [inline src/v0/sources/shopify/transform.js:205]: process1?

File: src/v0/sources/shopify/util.js [hunk 5]
```diff
@@ -169,7 +169,12 @@ const setAnonymousIdorUserIdFromDb = async (message, metricMetadata) => {
   }
   let anonymousIDfromDB;
   const executeStartTime = Date.now();
-  const redisVal = await RedisDB.getVal(`${cartToken}`);
+  let redisVal;
+  try {
+    redisVal = await RedisDB.getVal(`${cartToken}`);
+  } catch (e) {
+    redisVal = null;
```
Comment (koladilip) [inline src/v0/sources/shopify/util.js:176]: is there metrics for this already?

Comment (anantjain45823) [reply src/v0/sources/shopify/util.js:176]: No adding it

File: src/v0/sources/shopify/transform.js (hunk 2 above)
Comment (anantjain45823) [reply src/v0/sources/shopify/transform.js:184]: Removed

File: src/v0/sources/shopify/transform.js (hunk 4 above)
Comment (anantjain45823) [reply src/v0/sources/shopify/transform.js:205]: Removed

File: src/v0/sources/shopify/util.js (hunk 5 above)
Comment (anantjain45823) [reply src/v0/sources/shopify/util.js:176]: added

File: src/v0/sources/shopify/transform.js (hunk 3 above)
Comment (anantjain45823) [reply src/v0/sources/shopify/transform.js:189]: added

**Comment by koladilip** [comment]
LGTM

**Comment by sanpj2292** [comment]
[Non-blocking] I don't think `errorType: "Redis"` is making a lot of sense as the metric itself contains `.*redis.*`. Can we either get rid of the label or think about putting some other detail which thoroughly describes the problem ?

**Comment by anantjain45823** [comment]
> [Non-blocking] I don't think `errorType: "Redis"` is making a lot of sense as the metric itself contains `.*redis.*`. Can we either get rid of the label or think about putting some other detail which thoroughly describes the problem ?

I guess we can rename it to `RedisConnection`  as its the only error we are getting in dev. Otherwise we can use custom error. WDYT @sanpj2292

**Comment by sanpj2292** [comment]
Let's not have `errorType` now as the setting or getting, can be due to something else.
I'd say have one stat `redis_error` or something, where in your label can be something like `operation`. This label `operation` indicates when the error was thrown and the types of operation are `createConnection`, `quitConnection`, `get`, `set` etc,.

And as another label, we can have `key`, which makes a lot of sense to know if while getting or setting a particular key in Redis, we are facing some problem(I'm thinking this would not be that common a scenario)

This would give a lot clear picture when you're looking at stats around redis in transformer

@koladilip @utsabc  any thoughts here ?

File: src/v0/sources/shopify/util.js [hunk 6]
```diff
@@ -169,7 +169,16 @@ const setAnonymousIdorUserIdFromDb = async (message, metricMetadata) => {
...
   let anonymousIDfromDB;
   const executeStartTime = Date.now();
-  const redisVal = await RedisDB.getVal(`${cartToken}`);
+  let redisVal;
+  try {
+    redisVal = await RedisDB.getVal(`${cartToken}`);
+  } catch (e) {
+    stats.increment('shopify_events_lost_due_redis', {
+      ...metricMetadata,
+      timestamp: Date.now(),
```
[the remark this answers is not in the document]
Comment (sanpj2292) [reply src/v0/sources/shopify/util.js:178]: If you have any valid reason for needing `timestamp` as a label, please let us know

File: src/v0/sources/shopify/transform.js [hunk 7]
```diff
@@ -178,7 +179,15 @@ const isIdentifierEvent = (event) => {
 const processIdentifierEvent = async (event, metricMetadata) => {
   if (useRedisDatabase) {
     const setStartTime = Date.now();
-    await RedisDB.setVal(`${event.cartToken}`, { anonymousId: event.anonymousId });
+    try {
+      await RedisDB.setVal(`${event.cartToken}`, { anonymousId: event.anonymousId });
+    } catch (e) {
+      stats.increment('shopify_identifier_events_lost_due_redis', {
+        ...metricMetadata,
+        timestamp: Date.now(),
```
Comment (sanpj2292) [inline src/v0/sources/shopify/transform.js:187]: https://github.com/rudderlabs/rudder-transformer/pull/2107/files#r1182502988

File: src/v0/sources/shopify/util.js [hunk 8]
```diff
@@ -169,7 +169,16 @@ const setAnonymousIdorUserIdFromDb = async (message, metricMetadata) => {
...
+    });
+    redisVal = null;
+  }
   stats.timing('redis_get_latency', executeStartTime, {
```
Comment (sanpj2292) [inline src/v0/sources/shopify/util.js:182]: We should apply same logic mentioned below for `redis_*_latency` & change the stat name to `redis_latency` & add the operation as `get` or `set` or `delete` etc., which makes things very easy for us!
Please put some thought into this and let me know what you think
https://github.com/rudderlabs/rudder-transformer/pull/2107#issuecomment-1530963143

Comment (anantjain45823) [reply src/v0/sources/shopify/util.js:182]: Makes sense updating the changes

Comment (anantjain45823) [reply src/v0/sources/shopify/util.js]: Done.

File: src/v0/sources/shopify/util.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (anantjain45823) [reply src/v0/sources/shopify/util.js]: Done

File: src/v0/sources/shopify/transform.js (hunk 7 above)
Comment (anantjain45823) [reply src/v0/sources/shopify/transform.js]: Removed Date.now()

File: src/util/redisConnector.js [hunk 9]
```diff
@@ -93,20 +109,18 @@ const RedisDB = {
       const bytes = Buffer.byteLength(valueToStore, "utf-8");
       await this.client.setex(key, dataExpiry, valueToStore);
       stats.gauge('redis_set_val_size', bytes, {
-        timestamp: Date.now()
       });
     } catch (e) {
-      stats.increment('redis_set_val_error', {
-        error: e,
-        timestamp: Date.now()
+      stats.redisError({
```
Comment (sanpj2292) [inline src/util/redisConnector.js:114]: This should be `stats.increment("redis_error", {...tags})` & we shouldn't define a specific stat in `stats.js`

Comment (anantjain45823) [reply src/util/redisConnector.js]: fixed

**Review by sanpj2292 — DISMISSED** [review]

**Review by ItsSudip — DISMISSED** [review]

File: src/util/redisConnector.js [hunk 10]
```diff
@@ -66,14 +83,13 @@ const RedisDB = {
...
         });
       }
       return isJsonExpected ? JSON.parse(value) : value;
     } catch (e) {
-      stats.increment('redis_get_val_error', {
-        error: e,
-        timestamp: Date.now()
+      stats.increment("redis_error", {
+        operation: "get",
+        key: `${key}`
```
Comment (utsabc) [inline src/util/redisConnector.js:92]: Lets remove such high cardinality labels such as `key`

File: src/util/redisConnector.js [hunk 11]
```diff
@@ -93,20 +109,18 @@ const RedisDB = {
...
       stats.gauge('redis_set_val_size', bytes, {
-        timestamp: Date.now()
       });
     } catch (e) {
-      stats.increment('redis_set_val_error', {
-        error: e,
-        timestamp: Date.now()
+      stats.increment("redis_error", {
+        operation: "set",
+        key: `${key}`
```
Comment (utsabc) [inline src/util/redisConnector.js:116]: Same here

File: src/util/redisConnector.js (hunk 10 above)
Comment (anantjain45823) [reply src/util/redisConnector.js]: done

**Review by utsabc — APPROVED** [review]

File: src/v0/sources/shopify/util.js [hunk 12]
```diff
@@ -85,125 +79,111 @@ const extractEmailFromPayload = (event) => {
...
   });
-  if (redisVal !== null) {
-    anonymousIDfromDB = redisVal.anonymousId;
-  }
-  if (!isDefinedAndNotNull(anonymousIDfromDB)) {
-    /* this is for backward compatability when we don't have the redis mapping for older events
-    we will get anonymousIDFromDb as null so we will set UUID using the session Key */
-    stats.increment('shopify_no_anon_id_from_redis', {
+  if (anonymousId === null) {
+    stats.increment('shopify_redis_no_val', {
```
[the remark this answers is not in the document]
Comment (anantjain45823) [reply src/v0/sources/shopify/util.js:129]: this metric tells us that there is no data for the provided key. Hence Redis error can not be used

File: src/v0/sources/shopify/util.js [hunk 13]
```diff
@@ -85,125 +79,111 @@ const extractEmailFromPayload = (event) => {
...
-          : generateUUID(),
-      );
-      break;
-    default:
-      message.setProperty('anonymousId', generateUUID());
-      break;
+const getAnonymousId = (message) => {
+  const cartToken = getCartToken(message);
+  if (isDefinedAndNotNull(cartToken)) {
+    return v5(cartToken, v5.URL);
```
[the remark this answers is not in the document]
Comment (anantjain45823) [reply src/v0/sources/shopify/util.js:94]: v5 expects a valid UUID as a second parameter thats why I am using v5.url. 
For this it just gives Invalid UUID

File: src/v0/sources/shopify/util.js [hunk 14]
```diff
@@ -85,125 +79,111 @@ const extractEmailFromPayload = (event) => {
...
-    anonymousIDfromDB = sha256(cartToken).toString().substring(0, 36);
+    })
   }
-  message.setProperty('anonymousId', anonymousIDfromDB);
+  if (!isDefinedAndNotNull(anonymousId)) {
+    /* if redis does not have the mapping for cartToken as key (null) 
+      or redis is down(undefined)
+      we will set anonymousId as sha256(cartToken)
+     */
+    return v5(cartToken, v5.URL)
```
[the remark this answers is not in the document]
Comment (anantjain45823) [reply src/v0/sources/shopify/util.js:139]: same as above

File: src/v0/sources/shopify/util.js [hunk 15]
```diff
@@ -85,125 +79,111 @@ const extractEmailFromPayload = (event) => {
...
   }
-  let anonymousIDfromDB;
-  const executeStartTime = Date.now();
-  const redisVal = await RedisDB.getVal(`${cartToken}`);
-  stats.timing('redis_get_latency', executeStartTime, {
-    ...metricMetadata,
-  });
-  stats.increment('shopify_redis_get_data', {
+
+  stats.increment('shopify_redis_call', {
```
[the remark this answers is not in the document]
Comment (anantjain45823) [reply src/v0/sources/shopify/util.js]: done

File: src/v0/sources/shopify/util.js [hunk 16]
```diff
@@ -85,125 +79,111 @@ const extractEmailFromPayload = (event) => {
...
+}
+const checkAndUpdateCartItems = async (inputEvent, metricMetadata) => {
+  const cartToken = inputEvent.token || inputEvent.id;
+  let itemsHash;
+  try {
+    itemsHash = await RedisDB.getVal(cartToken, 'itemsHash');
+  } catch (e) {
+    // so if redis is down we will send the event to downstream destinations
+    console.log("Error in getting itemsHash from redis", e);
+    stats.increment('shopify_redis_call_failure', {
```
[the remark this answers is not in the document]
Comment (anantjain45823) [reply src/v0/sources/shopify/util.js]: done

File: src/v0/sources/shopify/transform.js [hunk 17]
```diff
@@ -172,14 +194,19 @@ const processEvent = async (inputEvent, metricMetadata) => {
...
+      event.cart?.line_items.length !== 0 ? sha256(event.line_items) : "0"];
+    try {
+      await RedisDB.setVal(`${event.cartToken}`, value);
+    } catch (e) {
+      stats.increment('shopify_redis_call_failure', {
+        type: "set",
+        ...metricMetadata,
+      });
+    }
+    stats.increment('shopify_redis_call', {
```
[the remark this answers is not in the document]
Comment (anantjain45823) [reply src/v0/sources/shopify/transform.js]: done

File: src/v0/sources/shopify/transform.js [hunk 18]
```diff
@@ -172,14 +194,19 @@ const processEvent = async (inputEvent, metricMetadata) => {
...
-    stats.timing('redis_set_latency', setStartTime, {
-      ...metricMetadata,
-    });
-    stats.increment('shopify_redis_set_anonymousId', {
+    const value = ["anonymousId", event.anonymousId, "itemsHash",
+      event.cart?.line_items.length !== 0 ? sha256(event.line_items) : "0"];
+    try {
+      await RedisDB.setVal(`${event.cartToken}`, value);
+    } catch (e) {
+      stats.increment('shopify_redis_call_failure', {
```
[the remark this answers is not in the document]
Comment (anantjain45823) [reply src/v0/sources/shopify/transform.js]: done

File: src/util/redis/redisConnector.js [hunk 19]
```diff
@@ -0,0 +1,153 @@
...
+    } catch (e) {
+      stats.increment("redis_error", {
+        operation: "set"
+      });
+      throw new RedisError(`Error setting value in Redis due ${e}`);
+    }
+  },
+  async disconnect() {
+    if (process.env.USE_REDIS_DB && process.env.USE_REDIS_DB !== 'false') {
+      stats.increment('redis_graceful_shutdown', {
```
[the remark this answers is not in the document]
Comment (anantjain45823) [reply src/util/redis/redisConnector.js]: done

File: src/util/redis/redisConnector.js [hunk 20]
```diff
@@ -0,0 +1,153 @@
...
+  async getVal(hashKey, key, isObjExpected = true) {
+    try {
+      await this.checkAndConnectConnection(); // check if redis is connected and if not, connect
+      let value;
+      if (isObjExpected === true) {
+        value = await this.client.hmget(hashKey, key);
+        try {
+          value = JSON.parse(value);
+        } catch (e) {
+          // do nothing
```
[the remark this answers is not in the document]
Comment (anantjain45823) [reply src/util/redis/redisConnector.js:91]: No thats not how it works.
Value can either be a string or an object which we don't know in advance. So we are trying to parse it as an object and if it throws an error that means it is string and we dont have to parse so we are just passing it as it is

File: src/util/redis/redisConnector.js [hunk 21]
```diff
@@ -0,0 +1,153 @@
...
+          }
+          stats.increment("redis_error", {
+            operation: 'redis_down',
+          });
+          log.error(`Redis is down at ${this.host}:${this.port}`);
+          return false; // stop retrying
+        },
+      });
+      this.client.on('ready', () => {
+        stats.increment('redis_ready', {
```
[the remark this answers is not in the document]
Comment (anantjain45823) [reply src/util/redis/redisConnector.js]: Not much. Removed as we are already logging it below it

**Comment by koladilip** [comment]
@anantjain45823 can you resolve these conflicts?

File: src/util/redis/redisConnector.js [hunk 22]
```diff
@@ -0,0 +1,158 @@
...
+    try {
+      await this.checkAndConnectConnection(); // check if redis is connected and if not, connect
+      let value;
+      let redisVal;
+      if (isObjExpected === true) {
+        redisVal = await this.client.hmget(hashKey, key);
+        if (isDefinedAndNotNull(redisVal)) {
+          value = redisVal.map(element => {
+            try {
+              return JSON.parse(element);
```
Comment (sanpj2292) [inline src/util/redis/redisConnector.js:93]: [Opt.] Can we use `isObject` in our `v0/util/index.js` ? Instead having a try-catch ?!

Also in this function would we also get array type results ?!

**Comment by anantjain45823** [comment]
> @anantjain45823 can you resolve these conflicts?

Conflicts resolved

File: src/util/redis/redisConnector.js (hunk 22 above)
Comment (sanpj2292) [reply src/util/redis/redisConnector.js]: Got it 👍🏼

**Review by sanpj2292 — APPROVED** [review]

File: src/v0/sources/shopify/util.js [hunk 23]
```diff
@@ -37,7 +37,12 @@ const getShopifyTopic = (event) => {
   }
   return topic[0];
 };
-
+const getHashLineItems = (cart) => {
+  if (cart && cart?.line_items && cart.line_items.length > 0) {
+    return sha256(JSON.stringify(cart.line_items));
+  }
+  return "0";
```
Comment (koladilip) [inline src/v0/sources/shopify/util.js:44]: should we use - "EMPTY" more descriptive right?

**Review by koladilip — COMMENTED** [review]
LGTM

File: src/v0/sources/shopify/util.js (hunk 23 above)
Comment (anantjain45823) [reply src/v0/sources/shopify/util.js:44]: Yes. This 0 was result of design discussion. Any how `EMPTY` would make much more sense. Changing it

Comment (anantjain45823) [reply src/v0/sources/shopify/util.js]: Done

File: src/util/redis/redisConnector.js [hunk 24]
```diff
@@ -101,6 +101,7 @@ const RedisDB = {
       stats.increment("redis_error", {
         operation: "get"
       });
+      log.error(`Error getting value from Redis: ${e}`);
```
Comment (koladilip) [inline src/util/redis/redisConnector.js:104]: isn't this is logged in versioned router?

Comment (anantjain45823) [reply src/util/redis/redisConnector.js:104]: No we are catching this error at destination level only

**Review by sanpj2292 — APPROVED** [review]

**Review by sanpj2292 — APPROVED** [review]

**Pull request merged**